GVFS/GVFS.Common/Tracing/QueuedPipeStringWriter.cs

```diff
@@ -0,0 +1,211 @@
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.IO.Pipes;
+using System.Text;
+using System.Threading;
+
+namespace GVFS.Common.Tracing
+{
+    public class QueuedPipeStringWriter : IDisposable
+    {
+        private static readonly TimeSpan PollingInterval = TimeSpan.FromSeconds(30);
+
+        private readonly Func<NamedPipeClientStream> pipeFactory;
+        private readonly ConcurrentQueue<string> queue = new ConcurrentQueue<string>();
+        private readonly AutoResetEvent queuePulse = new AutoResetEvent(initialState: false);
+
+        private Thread writerThread;
+        private NamedPipeClientStream pipeClient;
+        private bool isRunning;
+        private bool isDisposed;
+
+        public QueuedPipeStringWriter(Func<NamedPipeClientStream> pipeFactory)
+        {
+            this.pipeFactory = pipeFactory;
+        }
+
+        public event EventHandler<QueuedPipeStringWriterStateEventArgs> StateChanged;
+
+        public enum QueuedPipeStringWriterState
+        {
+            Stopped,
+            Failing,
+            Healthy,
+        }
+
+        public void Start()
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeStringWriter));
+            }
+
+            if (this.isRunning)
+            {
+                return;
+            }
+
+            this.isRunning = true;
+
+            this.writerThread = new Thread(this.ThreadProc)
+            {
+                Name = nameof(QueuedPipeStringWriter),
+                IsBackground = true,
+            };
+
+            this.writerThread.Start();
+        }
+
+        public void Enqueue(string message)
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeStringWriter));
+            }
+
+            if (!this.isRunning)
+            {
+                throw new InvalidOperationException($"{nameof(QueuedPipeStringWriter)} is not started");
+            }
+
+            this.queue.Enqueue(message);
+
+            // Wake the queue draining thread if it is waiting
+            this.queuePulse.Set();
+        }
+
+        public void Stop()
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeStringWriter));
+            }
+
+            if (!this.isRunning)
+            {
+                return;
+            }
+
+            // Signal to the queue draining thread that it should drain once more and then terminate.
+            // Setting isRunning to false will additionally prevent any new messages from being enqueued.
+            this.isRunning = false;
+            this.queuePulse.Set();
+            this.writerThread.Join();
+
+            Debug.Assert(this.queue.IsEmpty, "Message queue should be empty after being stopped");
```

Comment:
If this is an actual error/bad condition, could we throw an exception instead of adding a `Debug.Assert`?

If this assert were to start firing I'm not sure that we'd catch it (given our release/validation process) before shipping VFS4G (in general the code relies on unit tests and exceptions rather than asserts to ensure that expected conditions are being enforced).

Reply:
Actually, this `Debug.Assert` could be removed completely if we switched to using `BlockingCollection` as this scenario would be impossible.

I'm happy to help if you've run into problems with `BlockingCollection`, as I've found it can really simply scenarios like this.